### PR TITLE
Improve Chip Performance

### PR DESCRIPTION
**Pull request by gspencergoog**
This improves Chip performance by removing a composite step for a shaped container that isn't needed: instead, I blend the desired color of the container with the material color, and then set that as the material color.

Also, this significantly reduces the complexity of the chip demo in the gallery, since having 20 chips on the page was just too many chips to display to have a reasonable frame rate.

This addresses #16378

### CODE REVIEW

**Review by HansMuller — APPROVED** [review]
I'd appreciate it if you'd replace the shape-changing icon button's icon to Icons.sentiment_very_satisfied; the other demos that support a shape change use that one.

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 1]
```diff
@@ -116,176 +12,75 @@ class ChipDemo extends StatefulWidget {
...
   @override
   Widget build(BuildContext context) {
-    final List<Widget> chips = _materials.map<Widget>((String name) {
-      return new Chip(
-        key: new ValueKey<String>(name),
-        backgroundColor: _nameToColor(name),
-        label: new Text(_capitalize(name)),
-        onDeleted: () {
+    final List<Widget> chips = <Widget>[
+      const SizedBox(height: 8.0, width: 0.0),
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Fixed (by removing it: it wasn't really needed).

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 2]
```diff
@@ -116,176 +12,75 @@ class ChipDemo extends StatefulWidget {
...
-            _selectedMaterial = value ? name : '';
+            _filterSelected = value;
           });
         },
-      );
-    }).toList();
+      ),
+      new ActionChip(
+        label: const Text('ActionChip'),
+        onPressed: () {},
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Done.  I had to reorganize a bit to make this work, but it is more satisfying.

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 3]
```diff
@@ -116,176 +12,75 @@ class ChipDemo extends StatefulWidget {
...
-        onPressed: () {
-          setState(() {
-            _selectedAction = name;
-          });
-        },
+    if (_showInputChip) {
+      chips.add(
+        new InputChip(
+          avatar: const CircleAvatar(
+              backgroundImage: const AssetImage(
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Fixed.

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 4]
```diff
@@ -303,17 +98,30 @@ class _ChipDemoState extends State<ChipDemo> {
...
-            ? theme.chipTheme.copyWith(
+            ? chipTheme.copyWith(
                 shape: new BeveledRectangleBorder(
-                side: const BorderSide(width: 0.66, style: BorderStyle.solid, color: Colors.grey),
+                side: const BorderSide(
+                  width: 0.66,
+                  style: BorderStyle.solid,
+                  color: Colors.grey,
+                ),
                 borderRadius: new BorderRadius.circular(10.0),
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: It's annoying that BorderRadius.circular can't be const...

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 5]
```diff
@@ -303,17 +98,30 @@ class _ChipDemoState extends State<ChipDemo> {
...
-      ),
-      floatingActionButton: new FloatingActionButton(
-        onPressed: () => setState(_reset),
-        child: const Icon(Icons.refresh),
+            : chipTheme,
+        child: new Center(
+          child: new Column(
+            children: chips
+                .map<Widget>(
+                  (Widget chip) => new Padding(
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: OK, I changed it back to one line. (I had it one line when I originally wrote it, and then formatted it later).

**Comment by gspencergoog** [comment]
Changed to the happy face!  PTAL

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 6]
```diff
@@ -4,106 +4,93 @@
...
+              },
+            );
+          },
+        ),
+      );
+    }
+
+    return new Column(
+      children: chips.map<Widget>(
+            (Widget chip) => new Padding(padding: const EdgeInsets.all(8.0), child: chip),
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/material/chip_demo.dart]: indent

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart [hunk 7]
```diff
@@ -4,106 +4,93 @@
...
-                  .map((Widget chip) => new Padding(
-                        padding: const EdgeInsets.all(4.0),
-                        child: chip,
-                      ))
-                  .toList(),
+      new ActionChip(
+        label: const Text('ActionChip'),
+        onPressed: () {
+          setState(() {
+            Scaffold.of(context).showSnackBar(const SnackBar(
```
Comment (HansMuller) [inline examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Is the setState() needed here? I wouldn't think so.

Also indentation. You could write this like:
```
Scaffold.of(context).showSnackBar(
  const SnackBar(
    content: const Text('Activate!', textAlign: TextAlign.center),
  ),
);
```

**Review by HansMuller — APPROVED** [review]
LGTM

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart (hunk 7 above)
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Yeah, not needed.

File: examples/flutter_gallery/lib/demo/material/chip_demo.dart (hunk 6 above)
Comment (gspencergoog) [reply examples/flutter_gallery/lib/demo/material/chip_demo.dart]: Fixed.

**Comment by Hixie** [comment]
Having 20 chips doesn't seem like an unreasonable number of chips, if we can't render 20 chips we have a bigger problem IMHO.

**Comment by Hixie** [comment]
How much does the performance improve if you make the Material go down the painting codepath instead of the layer codepath?

**Comment by gspencergoog** [comment]
I'm closing this PR until I get something more concrete in place for improving the performance.

For posterity, here are the performance graphs of the Chip in various configurations (from animating color of all the chips on the screen for 2 seconds and averaging the number of frames drawn.  There were nothing but chips on the screen):
chart

I'm going to look at turning off clipping in Material for Chips, because that is clearly the cause of a lot of performance loss.  Here's just a simple chip with only a label.  Replacing the Material with a shaped Container yields a lot of performance gain:
chart 1

**Pull request closed without merging**